runtime/parachains/src/paras.rs

```diff
@@ -0,0 +1,1103 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The paras module is responsible for storing data on parachains and parathreads.
+//!
+//! It tracks which paras are parachains, what their current head data is in
+//! this fork of the relay chain, what their validation code is, and what their past and upcoming
+//! validation code is.
+//!
+//! A para is not considered live until it is registered and activated in this module. Activation can
+//! only occur at session boundaries.
+
+use sp_std::prelude::*;
+use sp_std::marker::PhantomData;
+use sp_runtime::traits::One;
+use primitives::{
+	parachain::{ValidatorId, Id as ParaId, ValidationCode, HeadData},
+};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	traits::Get,
+	weights::Weight,
+};
+use codec::{Encode, Decode};
+use crate::configuration;
+
+#[cfg(feature = "std")]
+use serde::{Serialize, Deserialize};
+
+pub trait Trait: system::Trait + configuration::Trait { }
+
+/// Metadata used to track previous parachain validation code that we keep in
+/// the state.
+#[derive(Default, Encode, Decode)]
+#[cfg_attr(test, derive(Debug, Clone, PartialEq))]
+pub struct ParaPastCodeMeta<N> {
+	// Block numbers where the code was "technically" replaced and the block number at
+	// which the code was actually replaced. These can be used as indices
+	// into the `PastCode` map along with the `ParaId` to fetch the code itself.
+	upgrade_times: Vec<(N, N)>,
+	// This tracks the highest pruned code-replacement, if any.
+	last_pruned: Option<N>,
+}
+
+#[cfg_attr(test, derive(Debug, PartialEq))]
+enum UseCodeAt<N> {
+	// Use the current code.
+	Current,
+	// Use the code that was replaced at the given block number.
+	ReplacedAt(N),
```

Comment:
The semantics of `ReplacedAt` aren't obvious. If the code was replaced at block `n`, and I `UseCodeAt::ReplacedAt(n)`, do I get the code which was active on `n` or `n-1`?

Elsewhere, we distinguish between 

> Block numbers where the code was "technically" replaced and the block number at
> which the code was actually replaced. 

Which of those does this `N` refer to?

This might all be solvable with increased documentation.

Reply:
Yes, I can add more docs for that.  This is from the perspective of the parachain, so `expected_at`, not `actually_at`.

> The semantics of ReplacedAt aren't obvious. If the code was replaced at block n, and I UseCodeAt::ReplacedAt(n), do I get the code which was active on n or n-1?

Am I understanding you correctly? Your question is whether, if there's a replacement at block `n`, which code we should get for `validation_code_at(n)`. The answer is `ReplacedAt(n)`, because the code that was replaced at block n is the code that was used to validate block n. Block n can't have been validated by code that was only replaced at block n.

Reply:
(I added some more docs, but above)